### PR TITLE
BUGFIX: Fix workspace role assignment after migration, import and site creation

### PR DESCRIPTION
**Pull request by bwaidelich**
With #5146 a basic workspace access control was implemented. With that, users won't have access to workspaces unless they have a role assigned.

This can be achieved via

```shell
./flow workspace:assignrole
```

This bugfix makes sure, that the default behavior (users with the role `Neos.Neos:LivePublisher` can collaborate on the `live` workspace) is automatically ensured when creating/importing a site.

Furthermore, the migration command has been fixed to add metadata & role assignments even if no workspace title was set:

```shell
./flow migrateevents:migrateWorkspaceMetadataToWorkspaceService
```

Related: #4726

### CODE REVIEW

File: Neos.Neos/Classes/Command/CrCommandController.php [hunk 1]
```diff
@@ -114,6 +119,9 @@ public function importCommand(string $path, string $contentRepository = 'default
         $projectionService = $this->contentRepositoryRegistry->buildService($contentRepositoryId, $this->projectionReplayServiceFactory);
         $projectionService->replayAllProjections(CatchUpOptions::create());
 
+        $this->outputLine('Assigning live workspace role');
+        $this->workspaceService->assignWorkspaceRole($contentRepositoryId, WorkspaceName::forLive(), WorkspaceRoleAssignment::createForGroup('Neos.Neos:LivePublisher', WorkspaceRole::COLLABORATOR));
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Command/CrCommandController.php:123]: imo as hotfix we also should create a site entity for each imported site node here :)

File: Neos.Neos/Classes/Domain/Service/SiteService.php [hunk 2]
```diff
@@ -176,7 +182,7 @@ public function createSite(
 
         $siteServiceInternals = $this->contentRepositoryRegistry->buildService(
             $site->getConfiguration()->contentRepositoryId,
-            new SiteServiceInternalsFactory()
+            new SiteServiceInternalsFactory($this->workspaceService)
```
Comment (pKallert) [inline Neos.Neos/Classes/Domain/Service/SiteService.php:185]: With this change, creating sites stops working for me. Could you check to make sure that this does not break site creation?

**Review by pKallert — CHANGES_REQUESTED** [review]
Tested it locally and the live workspace shows up again 👍 
Site creation in UI and console with site:create stopped working for me with this PR, even after flushing all caches, this should be checked before merging

File: Neos.Neos/Classes/Domain/Service/SiteService.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Domain/Service/SiteService.php:185]: @pKallert what does "stops working" mean? do you get an exception?

Comment (bwaidelich) [reply Neos.Neos/Classes/Domain/Service/SiteService.php:185]: ah, nevermind, I can reproduce it, nasty proxies
> Neos\Neos\Domain\Service\SiteServiceInternalsFactory_Original::__construct(): Argument #1 ($workspaceService) must be of type Neos\Neos\Domain\Service\WorkspaceService, Neos\Flow\ObjectManagement\DependencyInjection\DependencyProxy given

Comment (pKallert) [reply Neos.Neos/Classes/Domain/Service/SiteService.php:185]: Bildschirmfoto 2024-10-11 um 15 30 28

File: Neos.Neos/Classes/Command/CrCommandController.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Command/CrCommandController.php:123]: I'll take care of that in a follow-up

File: Neos.Neos/Classes/Domain/Service/SiteService.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Domain/Service/SiteService.php:185]: can you check again?

Comment (pKallert) [reply Neos.Neos/Classes/Domain/Service/SiteService.php:185]: Works now, thank you 😄

**Review by pKallert — APPROVED** [review]

File: Neos.Neos/Classes/Command/CrCommandController.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Command/CrCommandController.php:123]: also thie needs to be try catched, as a workspace role might already be assigned if you run this import before and then run `cr:prune` as the pruning would not prune the assignments ...

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php [hunk 3]
```diff
@@ -498,31 +499,14 @@ public function migrateWorkspaceMetadataToWorkspaceService(\Closure $outputFn):
...
-                        'baseWorkspaceName' => null,
-                        'workspaceOwner' => null
-                    ];
-                    break;
                 case 'WorkspaceWasCreated':
                     $eventData = self::decodeEventPayload($eventEnvelope);
-                    if (!isset($eventData['workspaceTitle'])) {
-                        // without the field it's not a legacy workspace creation event
-                        continue 2;
-                    }
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:519]: without this statement

> 'Found %d legacy workspace events resulting in %d workspaces.

is wrong. .. also this is okay not to be there? We have runtime migration fallback? Why did we remove this check?

**Pull request merged**